packages/upstash-redis/src/index.ts

```diff
@@ -0,0 +1,209 @@
+  accountByUserIdPrefix?: string
+  emailKeyPrefix?: string
+  sessionKeyPrefix?: string
+  sessionByUserIdKeyPrefix?: string
+  userKeyPrefix?: string
+  verificationTokenKeyPrefix?: string
+}
+
+const defaultOptions = {
+  baseKeyPrefix: "",
+  accountKeyPrefix: "user:account:",
+  accountByUserIdPrefix: "user:account:by-user-id:",
+  emailKeyPrefix: "user:email:",
+  sessionKeyPrefix: "user:session:",
+  sessionByUserIdKeyPrefix: "user:session:by-user-id:",
+  userKeyPrefix: "user:",
+  verificationTokenKeyPrefix: "user:token:",
+}
+
+export default function UpstashRedisAdapter(
+  client: Upstash,
+  options: UpstashRedisAdapterOptions = {}
+): Adapter {
+  const mergedOptions = {
+    ...defaultOptions,
+    ...options,
+  }
+
+  const { baseKeyPrefix } = mergedOptions
+  const accountKeyPrefix = baseKeyPrefix + mergedOptions.accountKeyPrefix
+  const accountByUserIdPrefix =
+    baseKeyPrefix + mergedOptions.accountByUserIdPrefix
+  const emailKeyPrefix = baseKeyPrefix + mergedOptions.emailKeyPrefix
+  const sessionKeyPrefix = baseKeyPrefix + mergedOptions.sessionKeyPrefix
+  const sessionByUserIdKeyPrefix =
+    baseKeyPrefix + mergedOptions.sessionByUserIdKeyPrefix
+  const userKeyPrefix = baseKeyPrefix + mergedOptions.userKeyPrefix
+  const verificationTokenKeyPrefix =
+    baseKeyPrefix + mergedOptions.verificationTokenKeyPrefix
+
+  const reviveFromJson = (json: string) =>
+    JSON.parse(json, (key, value) =>
+      key === "emailVerified" || key === "expires" ? new Date(value) : value
+    )
+
+  const setObjectAsJson = async (key: string, obj: any) =>
+    await client.set(key, JSON.stringify(obj))
+
+  const setAccount = async (id: string, account: AdapterAccount) => {
+    const accountKey = accountKeyPrefix + id
+    await setObjectAsJson(accountKey, account)
+    await client.set(accountByUserIdPrefix + account.userId, accountKey)
+    return account
+  }
+
+  const getAccount = async (id: string) => {
+    const response = await client.get(accountKeyPrefix + id)
+    if (!response.data) return null
+    return reviveFromJson(response.data)
+  }
+
+  const setSession = async (id: string, session: AdapterSession) => {
+    const sessionKey = sessionKeyPrefix + id
+    await setObjectAsJson(sessionKey, session)
+    await client.set(sessionByUserIdKeyPrefix + session.userId, sessionKey)
+    return session
+  }
+
+  const getSession = async (id: string) => {
+    const response = await client.get(sessionKeyPrefix + id)
+    if (!response.data) return null
+    return reviveFromJson(response.data)
+  }
+
+  const setUser = async (id: string, user: AdapterUser) => {
+    await setObjectAsJson(userKeyPrefix + id, user)
+    await client.set(`${emailKeyPrefix}${user.email as string}`, id)
+    return user
+  }
+
+  const getUser = async (id: string) => {
+    const response = await client.get(userKeyPrefix + id)
+    if (!response.data) return null
+    return reviveFromJson(response.data)
+  }
+
+  const adapter: Adapter = {
```

Comment:
```suggestion
  const adapter: Adapter = {
```

this is unnecessary. just return the adapter object directly. You will be able to drop all the argument typings.

Please also use the `async methodName() {}` format for consistency with other adapters.